pkg/providers/cloudstack/cloudstack.go

```diff
@@ -717,15 +801,138 @@ func (p *cloudstackProvider) generateCAPISpecForCreate(ctx context.Context, clus
+		if err != nil {
+			return nil, nil, err
+		}
+		controlPlaneTemplateName = cp.Spec.MachineTemplate.InfrastructureRef.Name
+	} else {
+		controlPlaneTemplateName = common.CPMachineTemplateName(clusterName, p.templateBuilder.now)
+	}
+
+	previousWorkerNodeGroupConfigs := cluster.BuildMapForWorkerNodeGroupsByName(currentSpec.Cluster.Spec.WorkerNodeGroupConfigurations)
+
+	workloadTemplateNames := make(map[string]string, len(newClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
+	kubeadmconfigTemplateNames := make(map[string]string, len(newClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations))
+	for _, workerNodeGroupConfiguration := range newClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		needsNewWorkloadTemplate, err := p.needsNewMachineTemplate(ctx, workloadCluster, currentSpec, newClusterSpec, workerNodeGroupConfiguration, vdc, previousWorkerNodeGroupConfigs)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		needsNewKubeadmConfigTemplate, err := p.needsNewKubeadmConfigTemplate(workerNodeGroupConfiguration, previousWorkerNodeGroupConfigs)
+		if err != nil {
+			return nil, nil, err
+		}
+		if !needsNewKubeadmConfigTemplate {
+			mdName := machineDeploymentName(newClusterSpec.Cluster.Name, workerNodeGroupConfiguration.Name)
+			md, err := p.providerKubectlClient.GetMachineDeployment(ctx, mdName, executables.WithCluster(bootstrapCluster), executables.WithNamespace(constants.EksaSystemNamespace))
+			if err != nil {
+				return nil, nil, err
+			}
+			kubeadmconfigTemplateName = md.Spec.Template.Spec.Bootstrap.ConfigRef.Name
+			kubeadmconfigTemplateNames[workerNodeGroupConfiguration.Name] = kubeadmconfigTemplateName
+		} else {
+			kubeadmconfigTemplateName = common.KubeadmConfigTemplateName(clusterName, workerNodeGroupConfiguration.Name, p.templateBuilder.now)
+			kubeadmconfigTemplateNames[workerNodeGroupConfiguration.Name] = kubeadmconfigTemplateName
+		}
+
+		if !needsNewWorkloadTemplate {
+			mdName := machineDeploymentName(newClusterSpec.Cluster.Name, workerNodeGroupConfiguration.Name)
+			md, err := p.providerKubectlClient.GetMachineDeployment(ctx, mdName, executables.WithCluster(bootstrapCluster), executables.WithNamespace(constants.EksaSystemNamespace))
+			if err != nil {
+				return nil, nil, err
+			}
+			workloadTemplateName = md.Spec.Template.Spec.InfrastructureRef.Name
+			workloadTemplateNames[workerNodeGroupConfiguration.Name] = workloadTemplateName
+		} else {
+			workloadTemplateName = common.WorkerMachineTemplateName(clusterName, workerNodeGroupConfiguration.Name, p.templateBuilder.now)
+			workloadTemplateNames[workerNodeGroupConfiguration.Name] = workloadTemplateName
+		}
+		p.templateBuilder.WorkerNodeGroupMachineSpecs[workerNodeGroupConfiguration.MachineGroupRef.Name] = p.machineConfigs[workerNodeGroupConfiguration.MachineGroupRef.Name].Spec
```

Comment:
Why are we modifying the template builder? Doesn't it receive all the info it needs when we call `GenerateCAPISpecControlPlane` and `GenerateCAPISpecWorkers`

Reply:
@vivek-koppuru do you have any thoughts on why this is being set? It's taken from vsphere

Reply:
It seems like this is the only place we are telling the templateBuilder what the machine config map is. It seems like it would be better if we pass it into the function instead of set it on the object, but that might be better as a follow up.

Reply:
It seems like we are building `templateBuilder.WorkerNodeGroupMachineSpecs` in the provider constructor
I could be missing something but we either don't need this line or we need to improve the construction logic in the controller (and then skip this line).
It seems like we only need `machineConfigs` to "update" `templateBuilder.WorkerNodeGroupMachineSpecs` and that is already available in the constructor.

Reply:
I think we're building those based on the old machineconfigs fetched from kubectl, right? So if some values have changed and we need to re-build the capi spec, shouldn't we persist the newer input machine configs instead of the ones we've fetched?

Reply:
Are you saying `p.machineConfigs` is built from the old version of machine configs? I was under the impression `p.machineConfigs` are the new ones: they are injected in the provider constructor and they don't seem to be modified (except when setting default values).

As far if this is blocking, if we don't fully understand what this code is doing and why is here, then it is blocking. Once we understand it, we can look at the possible solutions and evaluate if they can be addressed separately.

Reply:
Created https://github.com/aws/eks-anywhere/issues/1772 to address the issues caused by this implementation